# Assignment 1

In [166]:
import numpy as np
import pandas as pd
import sklearn as sl
import matplotlib as mp
import scipy as cp
import seaborn as sb
from matplotlib import pyplot as plt
import statsmodels.api as sm
from statsmodels.tsa.stattools import acf
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression, Ridge, Lasso
from sklearn.metrics import f1_score, mean_absolute_error, mean_squared_error, mean_absolute_percentage_error
from math import sqrt
from sklearn import linear_model as lm
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer
from scipy.linalg import toeplitz

## Part A

1.	Download the “dataset_lm.csv” file from Canvas and upload it to Jupyter Notebook. 

In [167]:
df = pd.read_csv(r"C:\Users\johns\OneDrive\Desktop\MBAN Semester 2\Predictive Analytics\Assignment 1\dataset_lm.csv")

Get a quick overview of the data through the use of the head and info function

In [168]:
df.head()

Dependent Var  Explanatory Var #1  Explanatory Var #2  Explanatory Var #3  \
0      56.293458           13.698667           50.639873                   0   
1      58.473431            2.714725           65.845845                   1   
2      94.195330           11.618072           65.072497                   0   
3      29.074583            0.818623           45.408996                   1   
4      86.035569            9.077544           73.548021                   0   

   Explanatory Var #4  Explanatory Var #5  Explanatory Var #6  \
0          -18.568035           45.121911           11.412501   
1          -25.105932           47.190213           10.080280   
2           -7.897464           52.163036           11.057301   
3          -18.316132           54.356714            5.029029   
4          -19.204165           47.186807           12.128134   

   Explanatory Var #7  Explanatory Var #8  Explanatory Var #9  \
0           56.410757                   2          -12.281132   
1           65.383107                   3          -36.763585   
2           82.812717                   0          -15.733547   
3           48.812471                   1          -12.825591   
4           62.520911                   2          -13.804860   

   Explanatory Var #10  Explanatory Var #11  Explanatory Var #12  \
0            38.996909            -3.010548            49.195073   
1            51.654939             4.991111            45.591729   
2            48.913837            -2.457696            56.608806   
3            45.851732            14.974177            47.362594   
4            47.765904             9.593982            53.700562   

   Explanatory Var #13  Explanatory Var #14  Explanatory Var #15  
0                    0           -21.153143            46.919314  
1                    0            -6.474403            53.383508  
2                    0           -27.903299            48.515026  
3                    1           -10.064411            55.266254  
4                    0           -17.546302            48.150543

In [169]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

2.	Run the OLS model by using the dependent and explanatory variables in the dataset. 

Set y to the target variable Dependent Var

Set X all of the explanatory variables by setting it to the entire dataset except the Dependent Var

In [170]:
y = df['Dependent Var']
X = df.drop('Dependent Var', axis = 1)

First, we created a regular model using the X and y created above. However, in the notes the model says that it would center if a constant was added.

Next, we added a constant to the X value then created the updated model.

In [171]:
model_reg = sm.OLS(y,X).fit()

In [172]:
x_updated = sm.add_constant(X)
model_updated = sm.OLS(y,x_updated).fit()

3.	Show your summary table in Python and interpret your results in the summary report.

In [173]:
model_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Sat, 28 Oct 2023   Prob (F-statistic):                        0.00
Time:                        15:50:16   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Explanatory Var #1      1.3078      0.013    102.159      0.000       1.283       1.333
Explanatory Var #2      1.7667      0.009    203.101      0.000       1.750       1.784
Explanatory Var #3      6.6083      0.177     37.232      0.000       6.259       6.957
Explanatory Var #4      2.0725      0.011    189.896      0.000       2.051       2.094
Explanatory Var #5     -0.7776      0.011    -68.372      0.000      -0.800      -0.755
Explanatory Var #6      0.0124      0.008      1.595      0.111      -0.003       0.028
Explanatory Var #7      0.0286      0.006      4.890      0.000       0.017       0.040
Explanatory Var #8      0.3510      0.080      4.398      0.000       0.194       0.508
Explanatory Var #9     -0.0293      0.010     -2.838      0.005      -0.050      -0.009
Explanatory Var #10     0.1292      0.011     11.409      0.000       0.107       0.152
Explanatory Var #11     0.0145      0.013      1.152      0.250      -0.010       0.039
Explanatory Var #12     0.0752      0.009      8.698      0.000       0.058       0.092
Explanatory Var #13     0.7414      0.177      4.198      0.000       0.394       1.089
Explanatory Var #14    -0.0218      0.011     -1.974      0.049      -0.043   -8.61e-05
Explanatory Var #15     0.1212      0.012     10.324      0.000       0.098       0.144
==============================================================================
Omnibus:                        0.087   Durbin-Watson:                   1.875
Prob(Omnibus):                  0.957   Jarque-Bera (JB):                0.188
Skew:                          -0.004   Prob(JB):                        0.910
Kurtosis:                       2.897   Cond. No.                         287.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [174]:
model_updated.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 1.488e+30
Date:                Sat, 28 Oct 2023   Prob (F-statistic):               0.00
Time:                        15:50:16   Log-Likelihood:                 11995.
No. Observations:                 422   AIC:                        -2.396e+04
Df Residuals:                     406   BIC:                        -2.389e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  32.0000   9.77e-14   3.28e+14      0.000      32.000      32.000
Explanatory Var #1      1.3000   7.89e-16   1.65e+15      0.000       1.300       1.300
Explanatory Var #2      1.7000   5.73e-16   2.97e+15      0.000       1.700       1.700
Explanatory Var #3      6.2000    1.1e-14   5.64e+14      0.000       6.200       6.200
Explanatory Var #4      2.1000   6.77e-16    3.1e+15      0.000       2.100       2.100
Explanatory Var #5     -0.9000   7.94e-16  -1.13e+15      0.000      -0.900      -0.900
Explanatory Var #6   4.163e-16   4.79e-16      0.869      0.385   -5.25e-16    1.36e-15
Explanatory Var #7   1.277e-15    3.7e-16      3.450      0.001    5.49e-16       2e-15
Explanatory Var #8  -7.661e-15   5.03e-15     -1.523      0.128   -1.75e-14    2.23e-15
Explanatory Var #9    7.72e-16   6.43e-16      1.201      0.231   -4.92e-16    2.04e-15
Explanatory Var #10 -1.693e-15   8.01e-16     -2.113      0.035   -3.27e-15   -1.18e-16
Explanatory Var #11  2.404e-15   7.75e-16      3.103      0.002    8.81e-16    3.93e-15
Explanatory Var #12 -1.429e-15    5.8e-16     -2.465      0.014   -2.57e-15    -2.9e-16
Explanatory Var #13 -1.155e-14   1.11e-14     -1.039      0.299   -3.34e-14    1.03e-14
Explanatory Var #14 -2.998e-15   6.82e-16     -4.396      0.000   -4.34e-15   -1.66e-15
Explanatory Var #15 -3.664e-15   8.12e-16     -4.512      0.000   -5.26e-15   -2.07e-15
==============================================================================
Omnibus:                        0.418   Durbin-Watson:                   0.650
Prob(Omnibus):                  0.811   Jarque-Bera (JB):                0.241
Skew:                           0.007   Prob(JB):                        0.887
Kurtosis:                       3.116   Cond. No.                     2.55e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.55e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

lets remove values that have a p-value greater than 0.05 as this tells us these values are not significantly different from 0

I kept removing variables until I was left with no variables with a p-value greater than 0.05

In [175]:
# Removal 1
x_updated = x_updated.drop("Explanatory Var #6", axis = 1)
x_updated = x_updated.drop("Explanatory Var #8", axis = 1)
x_updated = x_updated.drop("Explanatory Var #9", axis = 1)
x_updated = x_updated.drop("Explanatory Var #13", axis = 1)

# Removal 2
x_updated = x_updated.drop("Explanatory Var #11", axis = 1)
x_updated = x_updated.drop("Explanatory Var #12", axis = 1)
x_updated = x_updated.drop("Explanatory Var #15", axis = 1)

# Removal 3
x_updated = x_updated.drop("Explanatory Var #7", axis = 1)
x_updated = x_updated.drop("Explanatory Var #10", axis = 1)
x_updated = x_updated.drop("Explanatory Var #14", axis = 1)

In [176]:
model_updated = sm.OLS(y,x_updated).fit()
model_updated.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 3.870e+31
Date:                Sat, 28 Oct 2023   Prob (F-statistic):               0.00
Time:                        15:50:16   Log-Likelihood:                 12446.
No. Observations:                 422   AIC:                        -2.488e+04
Df Residuals:                     416   BIC:                        -2.486e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 32.0000    1.8e-14   1.78e+15      0.000      32.000      32.000
Explanatory Var #1     1.3000   2.65e-16    4.9e+15      0.000       1.300       1.300
Explanatory Var #2     1.7000   1.93e-16    8.8e+15      0.000       1.700       1.700
Explanatory Var #3     6.2000   3.69e-15   1.68e+15      0.000       6.200       6.200
Explanatory Var #4     2.1000   2.29e-16   9.18e+15      0.000       2.100       2.100
Explanatory Var #5    -0.9000   2.65e-16  -3.39e+15      0.000      -0.900      -0.900
==============================================================================
Omnibus:                        3.824   Durbin-Watson:                   1.371
Prob(Omnibus):                  0.148   Jarque-Bera (JB):                2.806
Skew:                          -0.017   Prob(JB):                        0.246
Kurtosis:                       2.602   Cond. No.                         785.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-squared and adjusted R-squared: 1.000 meaning that the independent variables explain all of the variation in the dependent variable

F-Statistic: 3.870e+31 is a very high value meaning the model is statistically significant

Prob(F-statistic): 0 is a very low value meaning the overall model is statistically significant

Coefficients (Constant): 32 meaning when all other values have a value of 0, the dependent variable would have a value of 32

Coefficients (Explanatory Variables): All of the explanatory variables have unique coefficients that would affect the model in their own unique way

t-statistic: tells us how significant each of the explanantory variables are, the farther from 0 the more significant. All of our values have very high or very low t-statistic values.

p-value: tells us the probaility that coeficients are not significantly different from 0, we have only kept the variables that have had a p-value less than 0.05

## Part B

1.	Use error values from the OLS model to calculate their standard deviation and autocorrelation values for the first three lags. 

To get the error values we will first get the predicted values

Then subtract the predicted values from the actual values to get the errors

In [177]:
pred_vals = model_updated.predict(x_updated)
errors = y - pred_vals
errors

0      1.421085e-14
1     -2.842171e-14
2      7.105427e-14
3      3.552714e-14
4      2.842171e-14
           ...     
417   -1.421085e-14
418    4.263256e-14
419   -7.105427e-15
420    1.421085e-14
421    7.105427e-15
Length: 422, dtype: float64

We then used the standard deviation function (std) to calculate the standard deviation of the errors

In [178]:
st_dev = np.std(errors)
st_dev

3.154798854246503e-14

Calculate the autocorrelation for the first three values using the built in autocorrelation function (acf)

In [179]:
autocorrelation = acf(errors, nlags=3)

ac1 = autocorrelation[1]
ac2 = autocorrelation[2]
ac3 = autocorrelation[3]

print(ac1)
print(ac2)
print(ac3)

0.025730176015300463
0.0719423706866769
-0.030938373917480142


2.	Then, run the GLS model accordingly. 

We first must build the covariance matrix using the values we have generated

The first is the standard deviation of the errors which was generated above

Next we use the the number one then the three autocorrelation values and put that into a list together

We finally set all of the remaining values to zero which is 418 in this case

In [180]:
cov_matrix = st_dev**2*toeplitz(np.append([1, ac1, ac2, ac3], np.zeros(418)))

In [181]:
gls_results = sm.GLS(y, x_updated, cov_matrix).fit()

3.	Show your summary table in Python and interpret your results in the summary report.

In [182]:
print(gls_results.summary())

                            GLS Regression Results                            
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            GLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 2.671e+31
Date:                Sat, 28 Oct 2023   Prob (F-statistic):               0.00
Time:                        15:50:17   Log-Likelihood:                 12364.
No. Observations:                 422   AIC:                        -2.472e+04
Df Residuals:                     416   BIC:                        -2.469e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 32.0000   2.19

R-squared and adjusted R-squared: 1.000 meaning that the independent variables explain all of the variation in the dependent variable

F-Statistic: 2.671e+31 is a very high value meaning the model is statistically significant

Prob(F-statistic): 0 is a very low value meaning the overall model is statistically significant

Coefficients (Constant): 32 meaning when all other values have a value of 0, the dependent variable would have a value of 32

Coefficients (Explanatory Variables): All of the explanatory variables have unique coefficients that would affect the model in their own unique way

t-statistic: tells us how significant each of the explanantory variables are, the farther from 0 the more significant. All of our values have very high or very low t-statistic values

p-value: tells us the probaility that coeficients are not significantly different from 0, we have only kept the variables that have had a p-value less than 0.05

## Part C

1.	Split the dataset into two as the training and test sets (test size = 0.5). 

In [183]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

2.	Run the Lasso model with alpha=1 and estimate the coefficients using the training set. 

In [184]:
model_lasso = lm.Lasso(alpha=1).fit(X_train,y_train)
model_lasso.coef_

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

3.	Then, calculate the mean absolute percentage error using the test set. 

Method 1

use the test data to create predictions, then use the correct calculation to get the absolute percentage errors. Finally, take the mean of that to get the mean absolute percentage errors

In [185]:
predictions = model_lasso.predict(X_test)

In [186]:
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100

In [187]:
mape = absolute_percentage_errors.mean()    

In [188]:
print("Mean Absolute Percentage Error (MAPE):", mape)

Mean Absolute Percentage Error (MAPE): 4.432190198291583


Method 2

Use the built in function to calculate the mean absolute percentage error

In [189]:
mape = mean_absolute_percentage_error(y_test, predictions)
mape = mape*100
print("Mean Absolute Percentage Error (MAPE):", mape)

Mean Absolute Percentage Error (MAPE): 4.432190198291583


4.	Find an approximate value for alpha that minimizes the mean absolute percentage error.

In [190]:
alphas = [0.001, 0.01, 0.1, 1, 10, 100]
mape = []

In [191]:
for i in alphas:
    model_lasso = Lasso(alpha=i)
    model_lasso.fit(X_train, y_train)

    lasso_pred = model_lasso.predict(X_test)

    mean_absolute_percent = (mean_absolute_percentage_error(y_test, lasso_pred)) * 100
    mape.append(mean_absolute_percent)

    print(f"Alpha: {i} - Mean Absolute Percentage Error: {mean_absolute_percent}")

Alpha: 0.001 - Mean Absolute Percentage Error: 0.0044128318074118334
Alpha: 0.01 - Mean Absolute Percentage Error: 0.04429881163345532
Alpha: 0.1 - Mean Absolute Percentage Error: 0.4432318654957986
Alpha: 1 - Mean Absolute Percentage Error: 4.432190198291583
Alpha: 10 - Mean Absolute Percentage Error: 8.587094351590437
Alpha: 100 - Mean Absolute Percentage Error: 44.28770209373842


After testing all of the alpha values, 0.001 yields the lowest mean absolute percentage error value

## Part D

1.	Use the demand data given in the table and develop an appropriate forecasting model (i.e., the tailored regularization discussed in the class—see your slides for more info) that exploits the available information given in the table as much as possible.

Create a dataframe for the given demand data

In [192]:
data = pd.DataFrame({'Month':     [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25],
                    'Demand':     [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144],
                    'Adv_Demand': [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]
                    })

Set y to the target variable of Demand

Set X to the explanantory variables of Month and Advanced Demand

In [193]:
X = data[['Month', 'Adv_Demand']]
y = data['Demand']

Split the data into training and testing sets

We used the same test size as above

In [194]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

Begin hyper parameter tuning by testing alpha values for both models. We will test alpha values of 0.001, 0.01, 0.1, 1, and 10

In [195]:
alphas = [0.001, 0.01, 0.1, 1, 10, 100]
lasso_param_grid = {'alpha': alphas}
ridge_param_grid = {'alpha': alphas}

lasso_grid = GridSearchCV(Lasso(), lasso_param_grid, cv=5, scoring='neg_mean_squared_error')
ridge_grid = GridSearchCV(Ridge(), ridge_param_grid, cv=5, scoring='neg_mean_squared_error')

lasso_grid.fit(X_train, y_train)
ridge_grid.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': [0.001, 0.01, 0.1, 1, 10, 100]},
             scoring='neg_mean_squared_error')

Select the best parameters for each model

In [196]:
best_alpha_lasso = lasso_grid.best_params_['alpha']
best_alpha_ridge = ridge_grid.best_params_['alpha']

Fit the models based on the best parameters

In [197]:
lasso_model = Lasso(alpha=best_alpha_lasso)
ridge_model = Ridge(alpha=best_alpha_ridge)

lasso_model.fit(X_train, y_train)
ridge_model.fit(X_train, y_train)

Ridge(alpha=100)

Evaluate the models by getting the predictions and then calculating the root mean squared error

In [198]:
lasso_predictions = lasso_model.predict(X_test)
ridge_predictions = ridge_model.predict(X_test)

lasso_rmse = sqrt(mean_squared_error(y_test, lasso_predictions))
ridge_rmse = sqrt(mean_squared_error(y_test, ridge_predictions))

print(f'Lasso RMSE: {lasso_rmse}')
print(f'Ridge RMSE: {ridge_rmse}')

Lasso RMSE: 17.31937837652882
Ridge RMSE: 17.295623361752114


2.	Interpret your results.

The RMSE value tells us that the better of the two models is the Ridge model as its RMSE value is 17.30 in comparison to that of the ridge model which has a RMSE value of 17.32

Both models do not have the best RMSE value as it is telling us that it can predict demand based on advanced demand within 17 of the actual demand. When dealing with demand values between 100 and 200 this is not very accurate. 